Sources/MapboxCoreNavigation/NavigationRouter.swift

```diff
@@ -0,0 +1,179 @@
+import Foundation
+import MapboxDirections
+import MapboxNavigationNative
+
+public class NavigationRouter {
+    
+    public typealias RequestId = UInt64
+    
+    public struct RoutingRequest {
+        public let id: RequestId
+        
+        // Intended retain cycle to prevent deallocating. `RoutingRequest` will be deleted once request completes.
+        let router: NavigationRouter
+        
+        public func cancel() {
+            router.finish(request: id)
+        }
+    }
+    
+    public enum RouterSource {
+        case online
+        case offline
+        case hybrid
+        
+        var nativeSource: RouterType {
+            switch self {
+            case .online:
+                return .online
+            case .offline:
+                return .onboard
+            case .hybrid:
+                return .hybrid
+            }
+        }
+    }
+    
+    public private(set) var activeRequests: [RequestId : RoutingRequest] = [:]
+    public private(set) var source: RouterSource
+    private var router: RouterInterface
+    private var settings: NavigationSettings
+    
+    public init(_ source: RouterSource = .hybrid, settings: NavigationSettings = .shared) {
+        self.source = source
+        self.settings = settings
+        
+        let factory = NativeHandlersFactory(tileStorePath: settings.tileStoreConfiguration.navigatorLocation.tileStoreURL?.path ?? "",
+                                            credentials: settings.directions.credentials)
+        self.router = MapboxNavigationNative.RouterFactory.build(for: source.nativeSource,
+                                                                 cache: factory.cacheHandle,
+                                                                 historyRecorder: factory.historyRecorder)
+    }
+    
+    deinit {
+        print(">>> dealloc")
+    }
+    
+    fileprivate func finish(request id: RequestId) {
+        router.cancelRequest(forToken: id)
+        activeRequests[id] = nil
+        print(">>> finished id: \(id)")
+    }
+    
+    public func doRequest<ResponseType: Codable>(options: DirectionsOptions,
+                                                 completion: @escaping (Result<ResponseType, DirectionsError>) -> Void) -> RequestId {
+        let directionsUri = settings.directions.url(forCalculating: options).absoluteString
+        var requestId: RequestId!
+        requestId = router.getRouteForDirectionsUri(directionsUri) { [weak self] (result, _) in // mind exposing response origin?
```

Comment:
What if instead of retaining `NavigationRouter` in `RoutingRequest` we retain it here with strong `self`? Have you tried that? Would it work?

Reply:
Retaining it here does not guarantee it'll work. It may happen that scheduling a response to worker thread takes longer then the rest of this method and the NN router is deinited and so will callback. I think we had this issue with tileset descriptors fetching method before.

Reply:
Also, it seems to me that tying it's lifetime to a request handle makes sense and is meant to be.